cluster/kubernetes/sync.go

```diff
@@ -2,31 +2,308 @@ package kubernetes
 
 import (
 	"bytes"
+	"crypto/sha1"
+	"encoding/hex"
 	"fmt"
 	"io"
 	"os/exec"
 	"sort"
 	"strings"
 	"time"
 
-	rest "k8s.io/client-go/rest"
-
 	"github.com/go-kit/kit/log"
+	"github.com/imdario/mergo"
 	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	rest "k8s.io/client-go/rest"
+
 	"github.com/weaveworks/flux"
 	"github.com/weaveworks/flux/cluster"
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+	"github.com/weaveworks/flux/policy"
+	"github.com/weaveworks/flux/resource"
+)
+
+const (
+	syncSetLabel       = kresource.PolicyPrefix + "sync-set"
+	checksumAnnotation = kresource.PolicyPrefix + "sync-checksum"
+
+	// The namespace to presume if something doesn't have one, and we
+	// haven't been told what to use as a fallback. This is what
+	// `kubectl` uses when there's no config setting the fallback
+	// namespace.
+	DefaultFallbackNamespace = "default"
 )
 
+// Sync takes a definition of what should be running in the cluster,
+// and attempts to make the cluster conform. An error return does not
+// necessarily indicate complete failure; some resources may succeed
+// in being synced, and some may fail (for example, they may be
+// malformed).
+func (c *Cluster) Sync(spec cluster.SyncSet) error {
+	logger := log.With(c.logger, "method", "Sync")
+
+	// Keep track of the checksum of each resource, so we can compare
+	// them during garbage collection.
+	checksums := map[string]string{}
+
+	// NB we get all resources, since we care about leaving unsynced,
+	// _ignored_ resources alone.
+	clusterResources, err := c.getResourcesBySelector("")
+	if err != nil {
+		return errors.Wrap(err, "collating resources in cluster for sync")
+	}
+
+	cs := makeChangeSet()
+	var errs cluster.SyncError
+	for _, res := range spec.Resources {
+		id := res.ResourceID().String()
+		// make a record of the checksum, whether we stage it to
+		// be applied or not, so that we don't delete it later.
+		csum := sha1.Sum(res.Bytes())
+		checkHex := hex.EncodeToString(csum[:])
+		checksums[id] = checkHex
+		if res.Policy().Has(policy.Ignore) {
+			logger.Log("info", "not applying resource; ignore annotation in file", "resource", res.ResourceID(), "source", res.Source())
+			continue
+		}
+		if cres, ok := clusterResources[id]; ok && cres.Policy().Has(policy.Ignore) {
```

Comment:
I am not entirely sure this is correct.

1. Git should the source of truth
2. Doesn't this prevent users from ever effectively removing the ignore policy on a git-controlled resource which was synced to the cluster? (i.e. Add resource file in  git -> Sync -> Mark as ignored -> Mark as not ignored -> Make changes in git)

Reply:
If it's marked as ignore, it won't be synced to the cluster -- so you can't end up with an unsyncable resource that way.

But you _can_ mark something as ignored directly in the cluster, i.e, with `kubectl annotate ...`. That means it won't be updated when syncing, and won't be garbage collected either. It is an escape hatch from gitops, basically. To return it to control of fluxd, you have to make sure the manifest (if there is one) doesn't have the annotation, and remove the annotation from the cluster resource.

Reply:
Ah, I see, an ignore annotation from git never reaches the cluster. That makes sense. I think that a comment clarifying what are we doing (checking for manually-ignored cluster resources) and why it's ok (flux will never apply ignore annotations from git), would be helpful

Reply:
Ah, I see, an ignore annotation from git never reaches the cluster. That makes sense. I think that a comment clarifying what are we doing (checking for manually-ignored cluster resources) and why it's ok (flux will never apply ignore annotations from git), would be helpful